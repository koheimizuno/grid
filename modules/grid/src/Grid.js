import React from 'react'
import styled from 'styled-components'
import { breakpoints } from 'grid-styled-constants'

export const toArray = n => Array.isArray(n) ? n : [ n ]
// todo: add px
export const percent = n => (typeof n !== 'number' || n > 1) ? n : `${n * 100}%`
export const media = size => `@media screen and (min-width:${size}em)`
export const width = breaks => props => props.width
  ? (
    toArray(props.width).map((n, i) => breaks[i]
      ? `${media(breaks[i])}{width:${percent(n)};}`
      : `width:${percent(n)};`
    ).join('\n')
  )
  : null


const align = props => props.align || 'top'

const Grid = styled.div`
  box-sizing: border-box;
  display: inline-block;
  vertical-align: ${align};
  width: 100%;
  ${width([ null, ...breakpoints ])}
`
  /* replace with box-styled
  padding-left: ${pad}px;
  padding-right: ${pad}px;
  */

// Move to separate package
export const Half = props => <Grid {...props} width={[ 1, 1/2 ]} />
export const Third = props => <Grid {...props} width={[ 1, 1/3 ]} />
export const Quarter = props => <Grid {...props} width={[ 1, 1/4 ]} />

const φ = (1 + Math.sqrt(5)) / 2
const ga = φ - 1
const gb = 1 - ga

export const GoldenA = props => <Grid {...props} width={[ 1, ga ]} />
export const GoldenB = props => <Grid {...props} width={[ 1, gb ]} />

export default Grid

