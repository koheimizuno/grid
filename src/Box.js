import React from 'react'
import styled from 'styled-components'
import { space, width, responsiveStyle } from 'styled-system'
import { number, string, array, oneOfType } from 'prop-types'
import tag from 'tag-hoc'
import propTypes from './propTypes'
import removeProps from './remove-props'

export const flex = responsiveStyle('flex')
export const order = responsiveStyle('order')

const Tag = tag(removeProps)
const Base = Tag('div')

const Box = styled(Base)([],
  { boxSizing: 'border-box' },
  width,
  space,
  flex,
  order
)
Box.displayName = 'Box'

const responsivePropType = oneOfType([
  number,
  string,
  array
])

Box.propTypes = Object.assign({}, propTypes, {
  flex: responsivePropType,
  order: responsivePropType
})

export default Box
