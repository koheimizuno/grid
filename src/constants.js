export const breakpoints = [
  40,
  52,
  64
]

export const space = [
  0,
  8,
  16,
  32,
  64
]

export default {
  breakpoints,
  space
}
