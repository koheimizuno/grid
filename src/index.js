export { default as Box } from './Box'
export { default as Flex } from './Flex'
export { default as theme } from './theme'

// for backwards compatibility
export { default as div } from 'clean-tag'
