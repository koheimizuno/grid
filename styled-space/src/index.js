import React from 'react'
import styled from 'styled-components'
import { space } from 'styled-system'

const classnames = (...args) => args.join(' ')
const getClassName = el => (el.props && el.props.className) || ''

export class StyledChildren extends React.Component {
  render () {
    const { className, children } = this.props
    const styledChildren = React.Children.toArray(children)
      .map(child => React.cloneElement(child, {
        className: classnames(getClassName(child), className)
      }))

    return (
      <React.Fragment>
        {styledChildren}
      </React.Fragment>
    )
  }
}

const Space = styled(StyledChildren)([], space)

Space.propTypes = space.propTypes

export default Space
