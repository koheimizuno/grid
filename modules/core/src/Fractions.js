import React from 'react'
import Grid from 'grid-styled-grid'

export const Half = props => <Grid {...props} width={[ 1, 1/2 ]} />
export const Third = props => <Grid {...props} width={[ 1, 1/3 ]} />
export const Quarter = props => <Grid {...props} width={[ 1, 1/4 ]} />

const Fractions = {
  Half,
  Third,
  Quarter
}

export default Fractions
