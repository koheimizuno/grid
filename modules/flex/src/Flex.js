import React from 'react'
import styled from 'styled-components'
import Box from 'box-styled'

// hoc to remove unwanted attributes
// const hoc = Comp => ({ width, ...props }) => <Comp {...props} w={width} />

const align = props => props.align ? `align-items:${props.align};` : null
const justify = props => props.justify ? `justify-content:${props.justify};` : null
const order = props => props.order ? `order:${props.order};` : null
const wrap = props => props.wrap ? `flex-wrap:wrap;` : null
const column = props => props.column ? `flex-direction:column;` : null

const Flex = styled(Box)`
  display: flex;
  ${align}
  ${justify}
  ${order}
  ${wrap}
  ${column}
`

export default Flex
