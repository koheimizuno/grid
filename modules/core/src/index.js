import Box from 'box-styled'
import Flex from 'flex-styled'
import Grid from 'grid-styled-grid'

import Golden from './Golden'
import { Half, Third, Quarter } from './Fractions'

export { default as Box } from 'box-styled'
export { default as Flex } from 'flex-styled'
export { default as Grid } from 'grid-styled-grid'

export { default as Golden } from './Golden'
export { Half, Third, Quarter } from './Fractions'

export default {
  Box,
  Flex,
  Grid,
  Golden,
  Half,
  Third,
  Quarter
}
