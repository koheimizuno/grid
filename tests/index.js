import React from 'react'
import { create as render } from 'react-test-renderer'
import { Box, Flex } from '../src'
import 'jest-styled-components'

const renderJSON = el => render(el).toJSON()

// Box
test('Box renders', () => {
  const json = renderJSON(<Box m={2} px={3} />)
  expect(json).toMatchSnapshot()
})

test('Box renders with props', () => {
  const json = renderJSON(<Box
    m={[ 1, 2 ]}
    px={[ 1, 2 ]}
    width={1}
    flex='1 1 auto'
    alignSelf='flex-start'
  />)
  expect(json).toMatchSnapshot()
  expect(json).toHaveStyleRule('width', '100%')
  expect(json).toHaveStyleRule('flex', '1 1 auto')
  expect(json).toHaveStyleRule('align-self', 'flex-start')
  expect(json).toHaveStyleRule('margin', '4px')
})

// Flex
test('Flex renders', () => {
  const json = renderJSON(<Flex />)
  expect(json).toMatchSnapshot()
  expect(json).toHaveStyleRule('display', 'flex')
})

test('Flex renders with props', () => {
  const json = renderJSON(
    <Flex
      flexWrap='wrap'
      flexDirection='column'
      alignItems='center'
      justifyContent='space-between'
    />
  )
  expect(json).toMatchSnapshot()
  expect(json).toHaveStyleRule('flex-wrap', 'wrap')
  expect(json).toHaveStyleRule('flex-direction', 'column')
  expect(json).toHaveStyleRule('align-items', 'center')
  expect(json).toHaveStyleRule('justify-content', 'space-between')
})

test('Flex renders with flexDirection prop', () => {
  const json = renderJSON(
    <Flex
      flexDirection='column'
    />
  )
  expect(json).toMatchSnapshot()
  expect(json).toHaveStyleRule('flex-direction', 'column')
})

test('Flex renders with responsive props', () => {
  const json = renderJSON(
    <Flex
      flexWrap={[ 'wrap', 'nowrap' ]}
      flexDirection={[ 'column', 'row' ]}
      alignItems={[ 'stretch', 'center' ]}
      justifyContent={[ 'space-between', 'center' ]}
    />
  )
  expect(json).toMatchSnapshot()
})

test('Box removes grid-styled props', () => {
  const json = renderJSON(
    React.createElement(Box, {
      id: 'hi',
      width: .5,
      color: 'blue',
      fontSize: 4,
      flexWrap: 'wrap'
    })
  )
  expect(json.props.id).toBe('hi')
  expect(json.props.width).toBe(undefined)
  expect(json.props.color).toBe(undefined)
  expect(json.props.fontSize).toBe(undefined)
})

test('Box accepts an is prop to change elements', () => {
  const json = renderJSON(
    React.createElement(Box, {
      is: 'h2'
    })
  )
  expect(json.type).toBe('h2')
})

test('Box accepts a css prop', () => {
  const json = renderJSON(
    <Box
      css={{
        outline: '4px solid red'
      }}
    />
  )
  expect(json).toMatchSnapshot()
  expect(json).toHaveStyleRule('outline', '4px solid red')
})

test('Flex accepts an is prop and keeps styles', () => {
  const json = renderJSON(
    <Flex is='footer'
      px={3}
      py={4}
      color='tomato'
      alignItems='center'
      justifyContent='center'
    />
  )
  expect(json.type).toBe('footer')
  expect(json).toHaveStyleRule('padding-left', '16px')
  expect(json).toHaveStyleRule('padding-right', '16px')
  expect(json).toHaveStyleRule('padding-top', '32px')
  expect(json).toHaveStyleRule('padding-bottom', '32px')
  expect(json).toHaveStyleRule('color', 'tomato')
  expect(json).toHaveStyleRule('align-items', 'center')
  expect(json).toHaveStyleRule('justify-content', 'center')
})

test('Flex omits grid-styled props', () => {
  const json = renderJSON(
    <Flex is='footer'
      px={3}
      py={4}
      color='tomato'
      alignItems='center'
      justifyContent='center'
    />
  )
  expect(json.props.px).toBe(undefined)
  expect(json.props.py).toBe(undefined)
  expect(json.props.color).toBe(undefined)
  expect(json.props.alignItems).toBe(undefined)
  expect(json.props.justifyContent).toBe(undefined)
})
