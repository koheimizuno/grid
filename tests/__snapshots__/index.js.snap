// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Box accepts a css prop 1`] = `
.c0 {
  box-sizing: border-box;
  outline: 4px solid red;
}

<div
  className="c0"
/>
`;

exports[`Box renders 1`] = `
.c0 {
  box-sizing: border-box;
  margin: 8px;
  padding-left: 16px;
  padding-right: 16px;
}

<div
  className="c0"
/>
`;

exports[`Box renders with props 1`] = `
.c0 {
  box-sizing: border-box;
  margin: 4px;
  padding-left: 4px;
  padding-right: 4px;
  width: 100%;
  -webkit-flex: 1 1 auto;
  -ms-flex: 1 1 auto;
  flex: 1 1 auto;
  -webkit-align-self: flex-start;
  -ms-flex-item-align: start;
  align-self: flex-start;
}

@media screen and (min-width:40em) {
  .c0 {
    margin: 8px;
    padding-left: 8px;
    padding-right: 8px;
  }
}

<div
  className="c0"
  width={1}
/>
`;

exports[`Flex renders 1`] = `
.c0 {
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

<div
  className="c0"
/>
`;

exports[`Flex renders with flexDirection prop 1`] = `
.c0 {
  box-sizing: border-box;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

<div
  className="c0"
/>
`;

exports[`Flex renders with props 1`] = `
.c0 {
  box-sizing: border-box;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

<div
  className="c0"
/>
`;

exports[`Flex renders with responsive props 1`] = `
.c0 {
  box-sizing: border-box;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

@media screen and (min-width:40em) {
  .c0 {
    -webkit-flex-wrap: nowrap;
    -ms-flex-wrap: nowrap;
    flex-wrap: nowrap;
    -webkit-flex-direction: row;
    -ms-flex-direction: row;
    flex-direction: row;
    -webkit-align-items: center;
    -webkit-box-align: center;
    -ms-flex-align: center;
    align-items: center;
    -webkit-box-pack: center;
    -webkit-justify-content: center;
    -ms-flex-pack: center;
    justify-content: center;
  }
}

<div
  className="c0"
/>
`;
